nupic/research/frameworks/vernon/mixins/representation_overlap.py

```diff
@@ -0,0 +1,177 @@
+        self.logger.info(f"Tracking representation overlap of modules: {tracked_names}")
+
+        # The targets will be collected in `self.error_loss` in a 1:1 fashion
+        # to the tensors being collected by the hooks.
+        self.ro_targets = torch.tensor([]).long()
+
+    def process_ro_args(self, ro_args):
+
+        ro_args = deepcopy(ro_args)
+
+        # Collect information about which modules to apply hooks to
+        include_names = ro_args.pop("include_names", [])
+        include_modules = ro_args.pop("include_modules", [])
+        include_patterns = ro_args.pop("include_patterns", [])
+        filter_args = dict(
+            include_names=include_names,
+            include_modules=include_modules,
+            include_patterns=include_patterns,
+        )
+
+        # Others args
+        plot_freq = ro_args.get("plot_freq", 1)
+        plot_args = ro_args.get("plot_args", {})
+        max_samples = ro_args.get("max_samples_to_plot", 1000)
+
+        assert isinstance(plot_freq, int)
+        assert isinstance(plot_args, dict)
+        assert isinstance(max_samples, int)
+        assert plot_freq > 0
+        assert max_samples > 0
+
+        return plot_freq, filter_args, max_samples
+
+    def run_epoch(self):
+
+        # Run the epoch with tracking enabled.
+        with self.ro_hook:
+            results = super().run_epoch()
+
+        # The epoch was iterated in `run_epoch` so epoch 0 is really epoch 1 here.
+        iteration = self.current_epoch + 1
+
+        # Create visualization, and update results dict.
+        if iteration % self.ro_plot_freq == 0:
+
+            for name, _, activations in self.ro_hook.get_statistics():
+
+                # Representation overlap matrix
+                visual = plot_repr_overlap_matrix(activations, self.ro_targets)
+                results.update({f"repr_overlap_matrix/{name}": visual})
+
+                # Representation overlap distributions:
+                #  * inter-class pairs
+                #  * intra-class pairs
+                inter_ol, intra_ol = plot_repr_overlap_distributions(activations,
+                                                                     self.ro_targets)
+                results.update({f"repr_overlap_interclass/{name}": inter_ol})
+                results.update({f"repr_overlap_intraclass/{name}": intra_ol})
+
+        return results
+
+    def error_loss(self, output, target, reduction="mean"):
+        """
+        This computes the loss and then saves the targets computed on this loss. This
+        mixin assumes these targets correspond, in a 1:1 fashion, to the samples seen
+        in the forward pass.
+        """
+        loss = super().error_loss(output, target, reduction=reduction)
+        if self.ro_hook.tracking:
+
+            # Targets were initialized on the cpu which could differ from the
+            # targets collected during the forward pass.
+            self.ro_targets = self.ro_targets.to(target.device)
```

Comment:
Sidenote: I understand why you need the name `ro_targets`, as to not conflict with the other `self.targets` definition.  In your last PR, if you recall, I mentioned that it could be good to generalize the functionality so we didn't need to keep repeating mixins over and over. That would also help with this naming conflict.

If you want, it could be a good exercise for you try and make such a general mixin. It basically needs to pair a hook, the modules that hook will track, and the functions it will plot for each hook. Ideally, the mixin could take a list of hooks and plot for each hook. 

No pressure though. We can do without. I'd prefer you optimize for your advancing the research questions you're going down. However, if you keep finding yourself writing similar mixins, then we can discuss how this could be done; we're not too far from it already.

Reply:
I agree this would be useful, but I'll leave `ro_targets` as is for now, to keep it simple.